Disable building XPC services when launching WebKit processes as extensions

pvollan opened the pull request:
#### c0d19cb34e3a93e3e4bdba390db548dc5ab51fe9
<pre>
Disable building XPC services when launching WebKit processes as extensions
<a href="https://bugs.webkit.org/show_bug.cgi?id=267248">https://bugs.webkit.org/show_bug.cgi?id=267248</a>
<a href="https://rdar.apple.com/120677588">rdar://120677588</a>

Reviewed by Elliott Williams and Brent Fulgham.

Disable building XPC services when launching WebKit processes as extensions. Additonally, this patch
is changing the bundle ID of the extensions to match the XPC service bundle IDs. This is to ensure
that existing bundle ID checks will continue to work. Also disable fallback to XPC service launch,
since these are not being built anymore. Instead, a temporary fallback using the old bundle IDs is
added, since the bundle IDs are normally cached on the system until they are re-registered.

* Source/WebKit/Configurations/BaseExtension.xcconfig:
* Source/WebKit/Configurations/BaseXPCService.xcconfig:
* Source/WebKit/Configurations/GPUExtension.xcconfig:
* Source/WebKit/Configurations/GPUService.xcconfig:
* Source/WebKit/Configurations/NetworkService.xcconfig:
* Source/WebKit/Configurations/NetworkingExtension.xcconfig:
* Source/WebKit/Configurations/WebContentCaptivePortalExtension.xcconfig:
* Source/WebKit/Configurations/WebContentExtension.xcconfig:
* Source/WebKit/Configurations/WebContentService.xcconfig:
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
(WebKit::ProcessLauncher::setIsRetryingLaunch):
(WebKit::ProcessLauncher::isRetryingLaunch const):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::serviceNameAndIdentifier):
(WebKit::launchWithExtensionKit):
(WebKit::ProcessLauncher::launchProcess):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/Scripts/copy-webcontent-resources-to-private-headers.sh:

Canonical link: <a href="https://commits.webkit.org/272868@main">https://commits.webkit.org/272868@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/50a6df77ed887a09edf196e0be056fd7137dddf0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33444 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12216 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35361 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36071 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/30388 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14569 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9362 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/29512 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33919 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10305 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29854 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8998 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/33718 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/9118 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37399 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30374 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/30194 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/35244 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/9246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/7183 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/33115 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11000 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/7733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/9831 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4290 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9974 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/Configurations/BaseXPCService.xcconfig:
```diff
@@ -37,6 +37,7 @@ APP_ICON_Release = WebKit.icns;
 APP_ICON_Production = ;
 
 EXCLUDED_SOURCE_FILE_NAMES[sdk=embedded*] = $(EXCLUDED_SOURCE_FILE_NAMES_$(CONFIGURATION)) *.xib MediaFormatReader.bundle;
+EXCLUDED_SOURCE_FILE_NAMES_YES[sdk=iphone*17.4*] = *;
```
I believe this excludes all of the XPC build files for iPhoneOS and iPhoneSimulator....

brentfulgham commented on Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
```diff
@@ -160,6 +160,8 @@ static void launchWithExtensionKit(ProcessLauncher& processLauncher, ProcessLaun
         if (error) {
             NSLog(@"Error launching process %@ error %@", process, error);
             RELEASE_LOG_FAULT(Process, "Error launching process, descriptrion '%s', reason '%s'", String([error localizedDescription]).utf8().data(), String([error localizedFailureReason]).utf8().data());
+#if !PLATFORM(IOS)
```
But here, you will launch XPC for IOS_SIMULATOR. One of these isn't right, I think?

pvollan replied:
Ah, that's a good point! Fixed in latest patch.

Thanks for reviewing!

emw-apple replied:
Wait, so the plan is to only use extensions on iOS devices? Why?

pvollan replied:
The plan is to use it on all platforms, this will be removed once we enable it on the remaining platforms.

Thanks for reviewing!

emw-apple replied:
Yes, this is iOS-only, but we want to avoid building/launching the XPC service on simulators _and_ devices, right?

If so: `PLATFORM(IOS)` is correct (it's true for simulators and devices), and the xcconfig filters should be `sdk=iphone*x.y` like you originally had (to match `iphoneos` and `iphonesimulator` SDK names).

> But here, you will launch XPC for IOS_SIMULATOR.

Basically, I think this is wrong. PLATFORM(IOS) is `TARGET_OS_IOS`, which true for device and simulator variants.

brentfulgham reviewed (commented):
I think the build and launch rules are inconsistent. Can you confirm?

brentfulgham commented on Source/WebKit/Configurations/BaseXPCService.xcconfig:
```diff
@@ -37,6 +37,7 @@ APP_ICON_Release = WebKit.icns;
 APP_ICON_Production = ;
 
 EXCLUDED_SOURCE_FILE_NAMES[sdk=embedded*] = $(EXCLUDED_SOURCE_FILE_NAMES_$(CONFIGURATION)) *.xib MediaFormatReader.bundle;
+EXCLUDED_SOURCE_FILE_NAMES[sdk=iphoneos17.4*] = *;
```
These will need to change for future iPhone OS SDKs.

emw-apple replied:
Yeah, I would recommend inverting it. Set `EXCLUDED_SOURCE_FILE_NAMES = *` and then set `INCLUDED_SOURCE_FILE_NAMES[sdk=iphone*x.y] = *` for iOS 17.0 - 17.3.

pvollan replied:
Ah, good point! Will fix.

brentfulgham reviewed (approved):
I think you should have Elliott do final review for this, but otherwise it looks fine to me.

emw-apple commented on Source/WebKit/Configurations/GPUService.xcconfig:
```diff
@@ -25,6 +25,7 @@
 
 PRODUCT_NAME = com.apple.WebKit.GPU;
 PRODUCT_BUNDLE_IDENTIFIER = $(PRODUCT_NAME);
+PRODUCT_BUNDLE_IDENTIFIER[sdk=iphoneos17.4*] = $(PRODUCT_NAME)Service;
```
Is this needed? If setting EXCLUDED_SOURCE_FILE_NAMES does successfully prevent us from generating the product, I don't see why we would need to customize the bundle ID, because there will be no Info.plist to write it to.

pvollan replied:
Yes, this appears to be needed, since the Info.plist is still being created.

emw-apple commented on Source/WebKit/Configurations/BaseXPCService.xcconfig:
```diff
@@ -36,11 +36,27 @@ APP_ICON_Debug = WebKit.icns;
+EXCLUDED_SOURCE_FILE_NAMES[sdk=iphone*17.3*] = $(EXCLUDED_EMBEDDED);
+EXCLUDED_SOURCE_FILE_NAMES[sdk=iphone*18.0*] = $(EXCLUDED_EMBEDDED);
+
+CS_IDENTITY = $(CODE_SIGN_IDENTITY);
+CODE_SIGN_IDENTITY[sdk=iphone*] = ;
+CODE_SIGN_IDENTITY[sdk=iphone*17.0*] = $(CS_IDENTITY);
+CODE_SIGN_IDENTITY[sdk=iphone*17.1*] = $(CS_IDENTITY);
+CODE_SIGN_IDENTITY[sdk=iphone*17.2*] = $(CS_IDENTITY);
+CODE_SIGN_IDENTITY[sdk=iphone*17.3*] = $(CS_IDENTITY);
+CODE_SIGN_IDENTITY[sdk=iphone*18.0*] = $(CS_IDENTITY);
```
nit: We don't usually alias setting names like this in xcconfig. These settings become environment variables everywhere, so aliasing leads to more stuff being exported.

emw-apple replied:
Would also appreciate a comment explaining that we are disabling code signing since the bundle is empty on platforms that use ExtensionKit.

pvollan replied:
Ah, good point, will add comment.

Thanks for reviewing!

emw-apple commented on Source/WebKit/Scripts/copy-webcontent-resources-to-private-headers.sh:
```diff
@@ -39,7 +39,9 @@ PROCESSED_INFOPLIST="${BUILT_PRODUCTS_DIR}/${INFOPLIST_PATH}"
 COPIED_INFOPLIST="${WEB_CONTENT_RESOURCES_PATH}/Info-WebContent.plist"
 echo "Copying Info.plist from ${UNPROCESSED_INFOPLIST} to ${COPIED_INFOPLIST}"
-ditto "${UNPROCESSED_INFOPLIST}" "${COPIED_INFOPLIST}"
+if [ -e "${UNPROCESSED_INFOPLIST}" ]; then
+    ditto "${UNPROCESSED_INFOPLIST}" "${COPIED_INFOPLIST}"
+fi
```
Not having `UNPROCESSED_INFOPLIST` exist in some configurations may raise workspace validation warnings, since the plist is listed as an input dependency of this script phase. I don't think you should try and fix it now, but if this comes up in the future (especially as we update the iOS SDK on various build bots), we might need to switch this script phase to emit discovered dependencies.

pvollan committed:
Disable building XPC services when launching WebKit processes as extensions
https://bugs.webkit.org/show_bug.cgi?id=267248
rdar://120677588

Reviewed by Elliott Williams and Brent Fulgham.

Disable building XPC services when launching WebKit processes as extensions. Additonally, this patch
is changing the bundle ID of the extensions to match the XPC service bundle IDs. This is to ensure
that existing bundle ID checks will continue to work. Also disable fallback to XPC service launch,
since these are not being built anymore. Instead, a temporary fallback using the old bundle IDs is
added, since the bundle IDs are normally cached on the system until they are re-registered.

* Source/WebKit/Configurations/BaseExtension.xcconfig:
* Source/WebKit/Configurations/BaseXPCService.xcconfig:
* Source/WebKit/Configurations/GPUExtension.xcconfig:
* Source/WebKit/Configurations/GPUService.xcconfig:
* Source/WebKit/Configurations/NetworkService.xcconfig:
* Source/WebKit/Configurations/NetworkingExtension.xcconfig:
* Source/WebKit/Configurations/WebContentCaptivePortalExtension.xcconfig:
* Source/WebKit/Configurations/WebContentExtension.xcconfig:
* Source/WebKit/Configurations/WebContentService.xcconfig:
* Source/WebKit/UIProcess/Launcher/ProcessLauncher.h:
(WebKit::ProcessLauncher::setIsRetryingLaunch):
(WebKit::ProcessLauncher::isRetryingLaunch const):
* Source/WebKit/UIProcess/Launcher/cocoa/ProcessLauncherCocoa.mm:
(WebKit::serviceNameAndIdentifier):
(WebKit::launchWithExtensionKit):
(WebKit::ProcessLauncher::launchProcess):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/Scripts/copy-webcontent-resources-to-private-headers.sh:

Canonical link: https://commits.webkit.org/272868@main
